R/weight.R

```diff
@@ -59,3 +59,103 @@ setMethod("weightVariables", "VariableCatalog", function (x) {
         return(c())
     }
 })
+
+#' Generate a weight vector
+#'
+#' This function allows you to generate a weight vector by supplying a set of
+#' categorical variables and the target distribution for each of the variable's categories.
+#' For instance if you wanted to create a weight variable which equally weighted four categories stored
+#' in `ds$var` you would call `makeWeight(ds$var ~ c(25, 25, 25, 25), name = "weight")`.
+#'
+#' @param ...
+#' A series of expressions of the form `variable ~ target_weights`. The variable must
+#' be a categorical crunch variable, and the target weights must be a numeric vector whose
+#' length is equal to the number of categories contained in the variable, and whose sum is equal to 100.
+#'
+#' @param name
+#' The name of the resulting variable
+#' @rdname makeWeight
+#' @return
+#' A crunch Variable Definiton of the weight variable
+#' @export
+#'
+#' @examples
+#' \dontrun{
+#' mtcars$cyl <- as.factor(mtcars$cyl)
+#' mtcars$gear <- as.factor(mtcars$gear)
+#' ds <- newDataset(mtcars, "mtcars")
+#' ds$weight <- makeWeight(ds$cyl ~ c(30, 30, 40, 0), ds$gear ~ c(20, 20, 60, 0), name = "weight" )
+#' as.vector(ds$weight)
+#' }
+makeWeight <- function(..., name) {
+    expr_list <- list(...)
+    out <- list(
+        name = name,
+        derivation = list(
+            `function` = "rake",
+            args = lapply(expr_list, generateWeightEntry))
+        )
+    class(out) <- "VariableDefinition"
+    out
+}
+
+#' Generate entry for [makeWeight]
+#'
+#' Utility function to catch formula errors passed to [makeWeight], and to generate
+#' the appropriate entry.
+#'
+#' @param expr
+#' An expression
+#' @return
+#' A list of the variable id and the target weights.
+#'
+#' @keywords internal
+#' @examples
+#' \dontrun{
+#' validateWeightExpression(ds$var ~ c(10, 30, 60))
+#' }
+#'
+generateWeightEntry <- function(expr) {
+    formula <- try(as.formula(expr), silent = TRUE)
+    if (is.error(formula)) {
+        halt(dQuote(substitute(expr)), " is not a valid formula, use the form ds$var ~ c(10, 20, 30)")
+    }
+    var     <- eval(expr[[2]], environment(expr))
+    varname <- deparse(expr[[2]])
+    targets <- eval(expr[[3]])
+
+    if (!is.Categorical(var)) {
+        halt(varname, " is not a categorical crunch variable")
+    }
+
+    n_categories <- length(categories(var))
+
+    if (length(targets) > n_categories) {
+        halt("Number of targets does not match number of categories for ", varname)
+    }
+    if (!all(is.numeric(targets))) {
+        halt("Targets are not numeric for ", varname)
+    }
+    if (!(sum(targets) == 100 || sum(targets) == 1)) {
+        halt("Targets do not add up to 100% for ", varname)
+    }
+    if (sum(targets) != 1) {
+        targets <- targets / 100
+    }
+
```

Comment:
Are there any other bad-input conditions you might be concerned about? What if the targets are length 0? What if there are `NA`s in `targets`?

Reply:
Target of length 0 is going to be non-numeric so I think that one's caught. NA's produce an error, but I should probably just infer that they are zero.

Reply:
It's technically possible though less likely for a user to accidentally produce a zero-length numeric vector than include NAs:

```
> x <- numeric(0)
> is.numeric(x)
[1] TRUE
> length(x)
[1] 0
```

And I think you should raise an error if there are NA targets because it's probably not intentional.